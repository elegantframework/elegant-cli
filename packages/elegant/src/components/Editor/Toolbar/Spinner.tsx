import clsx from 'clsx'
import { HTMLProps, forwardRef } from 'react'

export const Spinner = forwardRef<HTMLDivElement, HTMLProps<HTMLDivElement>>(({ className, ...rest }, ref) => {
  const spinnerClass = clsx('animate-spin rounded-full border-2 border-current border-t-transparent h-4 w-4', className)

  return <div className={spinnerClass} ref={ref} {...rest} />
})

Spinner.displayName = 'Spinner'
