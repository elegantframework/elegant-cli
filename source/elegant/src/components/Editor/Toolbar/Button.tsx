import clsx from 'clsx';
import React from 'react';

export type ButtonVariant = 'primary' | 'secondary' | 'tertiary' | 'quaternary' | 'ghost'
export type ButtonSize = 'medium' | 'small' | 'icon' | 'iconSmall'

export type ButtonProps = {
  variant?: ButtonVariant
  active?: boolean
  activeClassname?: string
  buttonSize?: ButtonSize
} & React.ButtonHTMLAttributes<HTMLButtonElement>

export const Button = React.forwardRef<HTMLButtonElement, ButtonProps>(
  (
    { active, buttonSize = 'medium', children, disabled, variant = 'primary', className, activeClassname, ...rest },
    ref,
  ) => {
    const buttonClassName = clsx(
      'flex group items-center justify-center border border-transparent gap-2 text-sm font-semibold rounded-md disabled:opacity-50 whitespace-nowrap',

      variant === 'primary' &&
        clsx(
          'text-white bg-black border-black dark:text-black dark:bg-white dark:border-white',
          !disabled &&
            !active &&
            'hover:bg-neutral-800 active:bg-neutral-900 dark:hover:bg-neutral-200 dark:active:bg-neutral-300',
          active && clsx('bg-neutral-900 dark:bg-neutral-300', activeClassname),
        ),

      variant === 'secondary' &&
        clsx(
          'text-neutral-900 dark:text-white',
          !disabled &&
            !active &&
            'hover:bg-neutral-100 active:bg-neutral-200 dark:hover:bg-neutral-900 dark:active:bg-neutral-800',
          active && 'bg-neutral-200 dark:bg-neutral-800',
        ),

      variant === 'tertiary' &&
        clsx(
          'bg-neutral-50 text-neutral-900 dark:bg-neutral-900 dark:text-white dark:border-neutral-900',
          !disabled &&
            !active &&
            'hover:bg-neutral-100 active:bg-neutral-200 dark:hover:bg-neutral-800 dark:active:bg-neutral-700',
          active && clsx('bg-neutral-200 dark:bg-neutral-800', activeClassname),
        ),

      variant === 'ghost' &&
        clsx(
          'bg-transparent border-transparent text-neutral-500 dark:text-neutral-400',
          !disabled &&
            !active &&
            'hover:bg-black/5 hover:text-neutral-700 active:bg-black/10 active:text-neutral-800 dark:hover:bg-white/10 dark:hover:text-neutral-300 dark:active:text-neutral-200',
          active && clsx('bg-black/10 text-neutral-800 dark:bg-white/20 dark:text-neutral-200', activeClassname),
        ),

      buttonSize === 'medium' && 'py-2 px-3',
      buttonSize === 'small' && 'py-1 px-2',
      buttonSize === 'icon' && 'w-8 h-8',
      buttonSize === 'iconSmall' && 'w-6 h-6',

      className,
    )

    return (
      <button ref={ref} disabled={disabled} className={buttonClassName} {...rest}>
        {children}
      </button>
    )
  },
)

Button.displayName = 'Button'
