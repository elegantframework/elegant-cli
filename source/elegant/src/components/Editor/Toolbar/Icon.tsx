import clsx from 'clsx';
import { icons } from 'lucide-react';
import { memo } from 'react';

export type IconProps = {
  name: keyof typeof icons
  className?: string
  strokeWidth?: number
}

export const Icon = memo(({ name, className, strokeWidth }: IconProps) => {
  const IconComponent = icons[name]

  if (!IconComponent) {
    return null
  }

  return <IconComponent className={clsx('w-4 h-4', className)} strokeWidth={strokeWidth || 2.5} />
})

Icon.displayName = 'Icon'
