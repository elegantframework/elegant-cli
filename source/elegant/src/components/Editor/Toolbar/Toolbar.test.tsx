import React from "react";
import renderer from 'react-test-renderer';
import { Editor } from '@tiptap/react';
import StarterKit from "@tiptap/starter-kit";
import { Toolbar } from "./Toolbar";
import { Plus } from "lucide-react";

describe('Editor Toolbar component', () => {
    it('renders the toolbar as expected', () => {
        const editor = new Editor({
            extensions: [StarterKit]
        });

        const menu = renderer
        .create(
            <Toolbar.Button onClick={() => {
                console.log('hello world!')
            }}>
                <Plus className='w-4 h-4' strokeWidth={2.5}/>
            </Toolbar.Button>
        )
        .toJSON();
        expect(menu).toMatchSnapshot();
    });
});