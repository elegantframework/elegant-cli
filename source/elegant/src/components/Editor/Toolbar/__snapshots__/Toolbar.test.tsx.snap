// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Editor Toolbar component renders the toolbar as expected 1`] = `
<button
  className="flex group items-center justify-center border border-transparent gap-2 text-sm font-semibold rounded-md disabled:opacity-50 whitespace-nowrap bg-transparent border-transparent text-neutral-500 dark:text-neutral-400 hover:bg-black/5 hover:text-neutral-700 active:bg-black/10 active:text-neutral-800 dark:hover:bg-white/10 dark:hover:text-neutral-300 dark:active:text-neutral-200 w-8 h-8 gap-1 min-w-[2rem] px-2 w-auto"
  onClick={[Function]}
>
  <svg
    className="lucide lucide-plus w-4 h-4"
    fill="none"
    height={24}
    stroke="currentColor"
    strokeLinecap="round"
    strokeLinejoin="round"
    strokeWidth={2.5}
    viewBox="0 0 24 24"
    width={24}
    xmlns="http://www.w3.org/2000/svg"
  >
    <path
      d="M5 12h14"
    />
    <path
      d="M12 5v14"
    />
  </svg>
</button>
`;
